docs/developer/web.rst

```diff
@@ -0,0 +1,224 @@
+.. _developer_web:
+
+
+Renga Web API
+=============
+
+This page describes a set of rules and recommendations that a developer of
+any **Renga** related web service should follow.
+
+.. note::
+
+   This is not a formal specification.
+
+0. First steps
+--------------
+
+Make sure that your data model has started to stabilize. Focus on consistent
+and concrete naming of your *resources* and possible relations between them.
+
+
+1. Specification and Best Practicies
+------------------------------------
+
+The focus is on finding a right balance between standards and pragmatic API
+design which should serve its purpose.
+
+- `JSON API Specification`_ - a specification for building APIs in JSON
+- `Pragmatic RESTful API`_ - blog post on best practices for designing
+  a pragmatic RESTful API
+
+.. _`JSON API Specification`: http://jsonapi.org/format/
+.. _`Pragmatic RESTful API`: http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api
+
+There are some key principles that the developer of an API should aim for.
+
+2. HTTP verbs
+-------------
+
+Most of the *resources* should support the four basic functions of `CRUD`_:
+
+**Create**
+
+    ``POST`` method should be used when a request is non-idempotent (for
+    example a new resource is added to a database).
+
+**Read**
+
+    ``GET`` method should not change the state of the object itself.
+    Side effects are possible for endpoints that require authorization only,
+    but in any case these side effects should only create new objects used
+    for capturing the resource lineage.
+
+**Update**
+
+    ``PUT`` method should be used to replace a full state of an existing
+    object. To avoid `lost update <https://www.w3.org/1999/04/Editing/#3.1>`_,
+    each request **must** include an ``ETag`` of the resource it is
+    replacing. The ``ETag`` must be part of the ``If-Match`` header.
+
```

Comment:
Can you elaborate on how this "ETag" works concretely? Is this something which is in the headers of the request or does the back-end send them explicitely?

Reply:
@3C111 the backend should send them explicitly. There is a simple example https://www.w3.org/1999/04/Editing/new-entry.txt

Reply:
Specified in https://tools.ietf.org/html/rfc7232